wazo_test_helpers/asset_launching_test_case.py

```diff
@@ -346,25 +381,79 @@ def _maybe_dump_docker_logs(cls):
             logger.debug('Container logs dumped to %s', logfile.name)
 
     @staticmethod
-    def get_log_directory():
+    def get_log_directory() -> str:
         if not AssetLaunchingTestCase.log_dir:
-            char_set = string.ascii_lowercase
-            default_logging_dir = '/tmp/wazo-integration-{}'.format(
-                ''.join(random.choice(char_set) for _ in range(8))
+            random_strings = "".join(
+                random.choice(string.ascii_lowercase) for _ in range(8)
             )
             AssetLaunchingTestCase.log_dir = os.getenv(
-                'WAZO_TEST_DOCKER_LOGS_DIR', default_logging_dir
+                'WAZO_TEST_DOCKER_LOGS_DIR', f'/tmp/wazo-integration-{random_strings}'
             )
             if not os.path.exists(AssetLaunchingTestCase.log_dir):
                 os.makedirs(AssetLaunchingTestCase.log_dir, mode=0o755)
-        return AssetLaunchingTestCase.log_dir
+        return str(AssetLaunchingTestCase.log_dir)
 
 
-def _run_cmd(cmd, stderr=True):
-    logger.debug('%s', cmd)
-    stderr = subprocess.STDOUT if stderr else None
+class AssetLaunchingTestCase(AbstractAssetLaunchingHelper, unittest.TestCase):
+    @classmethod
+    def setUpClass(cls) -> None:
+        cls.launch_service_with_asset()
 
-    completed_process = subprocess.run(cmd, stdout=subprocess.PIPE, stderr=stderr)
+    @classmethod
+    def tearDownClass(cls) -> None:
+        cls.stop_service_with_asset()
+
+
+def make_asset_fixture(
+    asset_class: type[AbstractAssetLaunchingHelper],
+) -> Callable[[], Generator[type[AbstractAssetLaunchingHelper], None, None]]:
+    """
+    Helper to create asset loading fixtures for Pytest in conftest.py or test files.
+
+    Usage:
+    ```python
+    class MyAssetClass(AbstractAssetLaunchingHelper):
+        assets_root = Path(__file__).parent / '..' / '..' / 'assets'
+        asset = 'asset-name'
+        service = 'my-service'
+
+    # Note: assignment is important to be detected by pytest.
+    asset_name_asset = make_asset_fixture(MyAssetClass)
+
+    # For test methods
+    def test_my_test(asset_name_asset: MyAssetClass) -> None:
+        asset_name_asset.service_status()  # Containers are up, and you can call methods on
+
+    # For whole classes:
+    @pytest.mark.usefixtures('asset_name_asset')
+    class TestExample(TestCase):
+        def test_my_test(self) -> None:
+            self.asset_name_asset.service_status()  # Containers are up, and you can call methods
+    """
+    asset_name = asset_class.asset.replace('-', '_')
+
+    @pytest.fixture(name=f'{asset_name}_asset', scope='session')
```

Comment:
The misleading part is that we have this line first:
```python
whatever = make_asset_fixture(MyAssetClass)
```
Then a test:
```python
def test_my_test(asset_name_asset: MyAssetClass) -> None:
```
If the `whatever` and `asset_name_asset` don't match (like in the wazo-agid PR where we have `base` vs `base_asset`), it's a lot harder to understand what is happening the first time you read the code.
I would prefer an explicit version like:
```python
@pytest.fixture
def asset_name_asset():
    yield from make_fixture(MyAssetClass)
```
where the magic of the asset name discovery is kept to a minimum and largely covered by the documentation from `pytest`.

About the fixture scope, I don't have a good suggestion... Maybe [the automatic grouping of tests](https://docs.pytest.org/en/6.2.x/fixture.html#automatic-grouping-of-tests-by-fixture-instances) would automatically teardown the fixtures when they are not needed anymore?

Reply:
Or if we want to keep a one-liner:
```python
whatever = make_asset_fixture(MyAssetClass, name='asset_name_asset')
```

Reply:
Yeha, hence the "whatever" in the example, to indicate that it is not relevent. However, I do agree that the `yield from` is a much cleaner solution. Thanks!

Hmm, good point. It would need to be tested in a project with multiple "assets", but yeah, according to the doc it should automatically work since the cleanup happens as soon as the fixture is no longer used.